File: ts/test/core.test.js
/* Copyright (c) 2026 Richard Rodger, MIT License */
'use strict'

// Core pipeline tests: one parse yields diagnostics + semantic tokens
// + outline, against the real engine (A1-A6 stack) and the strict-JSON
// fixture grammar. Protocol-free: server.js is wiring over this.

const { describe, it } = require('node:test')
const assert = require('node:assert')
const path = require('path')

const { Tabnas } = require('@tabnas/parser')
const { json } = require(path.join(
  require.resolve('@tabnas/parser').replace(/dist[/\\].*$/, ''),
  'dist-test',
  'json-plugin',
))

const { Doc } = require('../src/documents')
const { SerialInstances } = require('../src/instances')
const core = require('../src/core')
const { Registry, normalize } = require('../src/registry')

const ENTRY = normalize({
  name: '@tabnas/json-fixture',
  languageId: 'jsonf',
  extensions: ['.jsonf'],
  grammarKind: 'data',
  lexStream: 'clean',
})

function makeStack() {
  const instances = new SerialInstances(() => {
    const tn = new Tabnas({
      plugins: [json],
      parse: { recover: { enabled: true } },
    })
    ENTRY._inst = tn
    return tn
  })
  const inst = instances.get(ENTRY)
  return { instances, inst }
}

describe('lsp-core', () => {
  it('clean parse: no diagnostics, tokens and outline present', () => {
    const { instances, inst } = makeStack()
    const doc = new Doc('file:///t.jsonf', 'jsonf', 1, '{"a":[1,2]}')
    const a = core.analyze(instances, inst, ENTRY, doc)
    assert.deepStrictEqual(a.diagnostics, [])
    assert.ok(0 < a.semanticTokens.data.length, 'semantic tokens emitted')
    assert.equal(a.outline.length, 1)
    assert.equal(a.outline[0].name, 'Object')
    assert.equal(a.outline[0].children.length, 1)
    assert.equal(a.outline[0].children[0].name, 'Array')
  })

  it('broken document: diagnostics with ranges, partial value', () => {
    const { instances, inst } = makeStack()
    const doc = new Doc('file:///t.jsonf', 'jsonf', 1, '{"a":true blah,"b":2}')
    const a = core.analyze(instances, inst, ENTRY, doc)
    assert.ok(1 <= a.diagnostics.length)
    const d = a.diagnostics[0]
    assert.equal(d.severity, 1)
    assert.equal(d.range.start.line, 0)
    assert.ok(10 <= d.range.start.character, 'range at the bad word: ' + JSON.stringify(d.range))
    assert.ok(d.codeDescription.href.includes('/errors/'))
    assert.equal(a.value.b, 2)
  })

  it('multi-error document produces multiple diagnostics', () => {
    const { instances, inst } = makeStack()
    const doc = new Doc(
      'file:///t.jsonf', 'jsonf', 1,
      '{"a":true blah,"b":false blah,"c":true blah}',
    )
    const a = core.analyze(instances, inst, ENTRY, doc)
    assert.ok(2 <= a.diagnostics.length, 'got ' + a.diagnostics.length)
  })

  it('semantic tokens use delta encoding over multiple lines', () => {
    const { instances, inst } = makeStack()
    const doc = new Doc('file:///t.jsonf', 'jsonf', 1, '{"a":1,\n"b":2}')
    const a = core.analyze(instances, inst, ENTRY, doc)
    const data = a.semanticTokens.data
    assert.equal(0, data.length % 5)
    // Some token starts a new line (deltaLine 1).
    let sawNewline = false
    for (let i = 0; i < data.length; i += 5) if (1 === data[i]) sawNewline = true
    assert.ok(sawNewline, 'delta encoding crossed the newline')
  })

  it('multiline tokens split into line-local spans', () => {
    // A block comment spanning lines must not emit one token whose
    // length crosses the line break: multiline semantic tokens are an
    // optional client capability, and unsplit ones mis-highlight.
    //
    // Built on the shared pure-data grammar (comments ON — the strict
    // JSON plugin fixture disallows them), the same grammar the Go
    // mirror of this test uses.
    const fs = require('fs')
    const spec = JSON.parse(fs.readFileSync(
      path.join(__dirname, '..', '..', 'test', 'fixtures', 'json-grammar.json'),
      'utf8'))
    const entry = normalize({ name: 'specf', languageId: 'specf', grammarKind: 'data' })
    const instances = new SerialInstances(() => {
      const tn = new Tabnas({ parse: { recover: { enabled: true } } })
      tn.grammar(spec)
      entry._inst = tn
      return tn
    })
    const inst = instances.get(entry)
    const doc = new Doc('file:///t.specf', 'specf', 1, '{"a":1,/*x\ny*/"b":2}')
    const a = core.analyze(instances, inst, entry, doc)
    const lines = doc.text.split('\n')
    let line = 0
    let char = 0
    let sawContinuation = false
    const data = a.semanticTokens.data
    for (let i = 0; i < data.length; i += 5) {
      if (0 < data[i]) {
        line += data[i]
        char = data[i + 1]
      } else {
        char += data[i + 1]
      }
      const len = data[i + 2]
      assert.ok(char + len <= lines[line].length,
        'token crosses its line: line ' + line + ' char ' + char + ' len ' + len)
      if (1 === line && 0 === char) sawContinuation = true
    }
    assert.ok(sawContinuation, 'no continuation span on the second line')
  })

  it('the exported Instances class delivers collector events', () => {
    // The one instance-management class routes the permanent mux
    // through the active collector; a dead variant whose mux never
    // fired once shipped here (review catch on #1) — analyze() with it
    // returned empty tokens and outline while parsing succeeded.
    const { Instances } = require('../src/instances')
    const entry = normalize({
      name: '@tabnas/json-fixture', languageId: 'jsonf2',
      grammarKind: 'data', lexStream: 'clean',
    })
    const instances = new Instances(() => {
      const tn = new Tabnas({
        plugins: [json],
        parse: { recover: { enabled: true } },
      })
      entry._inst = tn
      return tn
    })
    const inst = instances.get(entry)
    const doc = new Doc('file:///t.jsonf', 'jsonf2', 1, '{"a":[1,2]}')
    const a = core.analyze(instances, inst, entry, doc)
    assert.ok(0 < a.semanticTokens.data.length, 'no lex events reached the collector')
    assert.equal(a.outline.length, 1, 'no ruleDone events reached the collector')
  })

  it('completion offers the colon after a key', () => {
    const { inst } = makeStack()
    const doc = new Doc('file:///t.jsonf', 'jsonf', 1, '{"a"')
    const items = core.completion(inst, ENTRY, doc, { line: 0, character: 4 })
    assert.ok(items.some((i) => ':' === i.label), JSON.stringify(items))
  })

  it('completion works mid-edit, not just at error points', () => {
    // The engine's continuations primitive now answers for prefixes
    // that parse (parser#101), which is what an editor actually asks
    // about. Before that fix these returned nothing.
    const { inst } = makeStack()
    const afterColon = new Doc('file:///t.jsonf', 'jsonf', 1, '{"a":')
    const items = core.completion(inst, ENTRY, afterColon, { line: 0, character: 5 })
    assert.ok(0 < items.length, 'value starters offered after a colon')

    const afterComma = new Doc('file:///t.jsonf', 'jsonf', 1, '[1,')
    const items2 = core.completion(inst, ENTRY, afterComma, { line: 0, character: 3 })
    assert.ok(
      items2.some((i) => ']' === i.label),
      'closer offered after a separator: ' + JSON.stringify(items2.map((i) => i.label)),
    )
  })

  it('registry routing: languageId gate and extension fallback', () => {
    const reg = new Registry([
      { name: '@tabnas/toml', languageId: 'toml', extensions: ['.toml'] },
      { name: '@tabnas/hoover', languageId: 'hoover', pluginKind: 'modifier' },
      { name: '@tabnas/yaml', languageId: 'yaml', extensions: ['.yaml'], enabled: false },
    ])
    assert.equal(reg.resolve('toml', 'file:///x.toml').entry.languageId, 'toml')
    // Generic client id falls back to extension matching.
    assert.equal(reg.resolve('plaintext', 'file:///x.toml').entry.languageId, 'toml')
    // Disabled entries are not routed even by extension.
    assert.equal(reg.resolve('plaintext', 'file:///x.yaml').entry, null)
    // Modifiers are never routing targets.
    assert.equal(reg.resolve('hoover', 'file:///x.hoover').entry, null)
  })

  it('quarantine disables a repeatedly failing grammar', () => {
    const bad = normalize({ name: '@tabnas/bad', languageId: 'bad' })
    const instances = new SerialInstances(() => {
      throw new Error('boom')
    })
    for (let i = 0; i < 3; i++) {
      try {
        instances.get(bad)
      } catch (e) {
        // expected
      }
    }
    assert.equal(instances.get(bad), null, 'quarantined after repeated failures')
  })

  it('lexStream gate: speculative grammars get no semantic tokens', () => {
    const { instances, inst } = makeStack()
    const spec = Object.assign({}, ENTRY, { lexStream: 'speculative' })
    const doc = new Doc('file:///t.jsonf', 'jsonf', 1, '{"a":1}')
    const a = core.analyze(instances, inst, spec, doc)
    assert.equal(a.semanticTokens, null)
  })

  it('completion drops engine sentinels', () => {
    // The engine names #ZZ (end-of-source) as a legal continuation for
    // every prefix that parses — its way of saying the document is
    // already valid. It is not something a user types, so it must not
    // reach the item list; a complete document offers nothing at all
    // rather than an item labelled "#ZZ".
    const { inst } = makeStack()
    const done = new Doc('file:///t.jsonf', 'jsonf', 1, '{"a":1}')
    const items = core.completion(inst, ENTRY, done, { line: 0, character: 7 })
    assert.deepStrictEqual(items, [])

    // A mid-edit prefix keeps its real completions; only the sentinel
    // is removed.
    const mid = new Doc('file:///t.jsonf', 'jsonf', 1, '[1,')
    const items2 = core.completion(inst, ENTRY, mid, { line: 0, character: 3 })
    assert.ok(0 < items2.length, 'real completions survive')
    for (const i of items2) {
      assert.ok('#ZZ' !== i.detail, 'sentinel leaked: ' + JSON.stringify(items2))
    }
  })
})

File: go/core_test.go
/* Copyright (c) 2026 Richard Rodger, MIT License */

package lsp

// The Go mirror of ts/test/core.test.js, against the shared pure-data
// strict-JSON grammar (test/fixtures/json-grammar.json — the L2 lane
// both runtimes load, which is exactly what makes it the parity test
// grammar).

import (
	"os"
	"path/filepath"
	"testing"

	tabnas "github.com/tabnas/parser/go"
)

func fixture(t *testing.T, name string) []byte {
	t.Helper()
	b, err := os.ReadFile(filepath.Join("..", "test", "fixtures", name))
	if err != nil {
		t.Fatalf("fixture %s: %v", name, err)
	}
	return b
}

func makeStack(t *testing.T) (*Instances, *tabnas.Tabnas, *Entry) {
	t.Helper()
	entry, makeInst := EntryFromSpecJSON("jsonf", []string{".jsonf"}, fixture(t, "json-grammar.json"))
	instances := NewInstances(makeInst)
	inst, err := instances.Get(entry)
	if err != nil {
		t.Fatalf("instance: %v", err)
	}
	return instances, inst, entry
}

func doc(text string) *Doc {
	return &Doc{URI: "file:///t.jsonf", LanguageID: "jsonf", Version: 1, Text: text}
}

func TestCleanParse(t *testing.T) {
	instances, inst, entry := makeStack(t)
	a := Analyze(instances, inst, entry, doc(`{"a":[1,2]}`))
	if 0 != len(a.Diagnostics) {
		t.Fatalf("diagnostics on a clean parse: %+v", a.Diagnostics)
	}
	if nil == a.SemanticTokens || 0 == len(a.SemanticTokens.Data) {
		t.Fatal("no semantic tokens emitted")
	}
	if 1 != len(a.Outline) {
		t.Fatalf("outline roots = %d, want 1", len(a.Outline))
	}
	if "Object" != a.Outline[0].Name {
		t.Fatalf("root symbol = %s, want Object", a.Outline[0].Name)
	}
	if 1 != len(a.Outline[0].Children) || "Array" != a.Outline[0].Children[0].Name {
		t.Fatalf("child symbols = %+v, want one Array", a.Outline[0].Children)
	}
}

func TestBrokenDocumentDiagnostics(t *testing.T) {
	instances, inst, entry := makeStack(t)
	a := Analyze(instances, inst, entry, doc(`{"a":true blah,"b":2}`))
	if 1 > len(a.Diagnostics) {
		t.Fatal("no diagnostics for a broken document")
	}
	d := a.Diagnostics[0]
	if 1 != d.Severity {
		t.Fatalf("severity = %d", d.Severity)
	}
	if 0 != d.Range.Start.Line {
		t.Fatalf("line = %d", d.Range.Start.Line)
	}
	if 10 > d.Range.Start.Character {
		t.Fatalf("range should sit at the bad word, got %+v", d.Range)
	}
	if nil == d.CodeDescription {
		t.Fatal("no codeDescription link")
	}
}

func TestMultiErrorDiagnostics(t *testing.T) {
	instances, inst, entry := makeStack(t)
	a := Analyze(instances, inst, entry,
		doc(`{"a":true blah,"b":false blah,"c":true blah}`))
	if 2 > len(a.Diagnostics) {
		t.Fatalf("recovery should yield multiple diagnostics, got %d", len(a.Diagnostics))
	}
}

func TestSemanticTokensDeltaEncoding(t *testing.T) {
	instances, inst, entry := makeStack(t)
	a := Analyze(instances, inst, entry, doc("{\"a\":1,\n\"b\":2}"))
	data := a.SemanticTokens.Data
	if 0 != len(data)%5 {
		t.Fatalf("data length %d not a multiple of 5", len(data))
	}
	sawNewline := false
	for i := 0; i < len(data); i += 5 {
		if 1 == data[i] {
			sawNewline = true
		}
	}
	if !sawNewline {
		t.Fatal("delta encoding never crossed the newline")
	}
}

func TestLexStreamGate(t *testing.T) {
	instances, inst, entry := makeStack(t)
	spec := *entry
	spec.LexStream = "speculative"
	a := Analyze(instances, inst, &spec, doc(`{"a":1}`))
	if nil != a.SemanticTokens {
		t.Fatal("speculative entries must not get semantic tokens")
	}
}

func TestCompletionOffersColonAfterKey(t *testing.T) {
	instances, inst, _ := makeStack(t)
	items := Complete(instances, inst, doc(`{"a"`), Position{Line: 0, Character: 4})
	found := false
	for _, i := range items {
		if ":" == i.Label {
			found = true
		}
	}
	if !found {
		t.Fatalf("no ':' completion after a key: %+v", items)
	}
}

func TestCompletionMidEdit(t *testing.T) {
	instances, inst, _ := makeStack(t)
	items := Complete(instances, inst, doc(`{"a":`), Position{Line: 0, Character: 5})
	if 0 == len(items) {
		t.Fatal("no value starters offered after a colon")
	}
	items2 := Complete(instances, inst, doc(`[1,`), Position{Line: 0, Character: 3})
	found := false
	for _, i := range items2 {
		if "]" == i.Label {
			found = true
		}
	}
	if !found {
		t.Fatalf("no closer offered after a separator: %+v", items2)
	}
}

func TestCompletionDropsSentinels(t *testing.T) {
	instances, inst, _ := makeStack(t)
	items := Complete(instances, inst, doc(`{"a":1}`), Position{Line: 0, Character: 7})
	if 0 != len(items) {
		t.Fatalf("a complete document should offer nothing, got %+v", items)
	}
	items2 := Complete(instances, inst, doc(`[1,`), Position{Line: 0, Character: 3})
	if 0 == len(items2) {
		t.Fatal("real completions must survive the sentinel filter")
	}
	for _, i := range items2 {
		if "#ZZ" == i.Detail {
			t.Fatalf("sentinel leaked: %+v", items2)
		}
	}
}

func TestQuarantine(t *testing.T) {
	bad := &Entry{Name: "bad", LanguageID: "bad", Enabled: true}
	instances := NewInstances(func(e *Entry) (*tabnas.Tabnas, error) {
		return nil, os.ErrInvalid
	})
	for i := 0; i < QuarantineLimit; i++ {
		if _, err := instances.Get(bad); err == nil {
			t.Fatal("expected a load error")
		}
	}
	inst, err := instances.Get(bad)
	if nil != inst || nil != err {
		t.Fatalf("want quarantined (nil, nil), got (%v, %v)", inst, err)
	}
}

func TestRegistryRouting(t *testing.T) {
	reg := NewRegistry([]*Entry{
		{Name: "toml", LanguageID: "toml", Extensions: []string{".toml"}, Enabled: true},
		{Name: "hoover", LanguageID: "hoover", PluginKind: "modifier", Enabled: true},
		{Name: "yaml", LanguageID: "yaml", Extensions: []string{".yaml"}, Enabled: false},
	})
	if e := reg.Resolve("toml", "file:///x.toml").Entry; nil == e || "toml" != e.LanguageID {
		t.Fatal("languageId routing failed")
	}
	if e := reg.Resolve("plaintext", "file:///x.toml").Entry; nil == e || "toml" != e.LanguageID {
		t.Fatal("extension fallback failed")
	}
	if e := reg.Resolve("plaintext", "file:///x.yaml").Entry; nil != e {
		t.Fatal("disabled entries must not route")
	}
	if e := reg.Resolve("hoover", "file:///x.hoover").Entry; nil != e {
		t.Fatal("modifiers must never be routing targets")
	}
}

func TestAstralRangeConversion(t *testing.T) {
	// "€" is one UTF-16 unit; "𝄞" (U+1D11E) is TWO. Engine columns are
	// runes; the LSP range must count UTF-16 units.
	d := doc("\"𝄞x\" q")
	// Engine sees: col 1 '"', col 2 '𝄞', col 3 'x', col 4 '"', col 5
	// ' ', col 6 'q'. In UTF-16: '"'=0, '𝄞'=1..2, 'x'=3, '"'=4, ' '=5,
	// 'q'=6.
	p := d.PosFromEngine(1, 6)
	if 6 != p.Character {
		t.Fatalf("astral conversion off: got %d, want 6", p.Character)
	}
	r := d.RangeFrom(1, 2, 1) // the 𝄞 token itself, len 1 code point
	if 1 != r.Start.Character || 3 != r.End.Character {
		t.Fatalf("astral token range %+v, want chars 1..3", r)
	}
}

func TestMultilineTokensSplitPerLine(t *testing.T) {
	// Mirrors the TS case "multiline tokens split into line-local
	// spans": a block comment spanning lines must not emit one token
	// whose length crosses the line break.
	instances, inst, entry := makeStack(t)
	d := doc("{\"a\":1,/*x\ny*/\"b\":2}")
	a := Analyze(instances, inst, entry, d)
	lines := []string{"{\"a\":1,/*x", "y*/\"b\":2}"}
	line, char := 0, 0
	sawContinuation := false
	data := a.SemanticTokens.Data
	for i := 0; i+4 < len(data); i += 5 {
		if 0 < data[i] {
			line += data[i]
			char = data[i+1]
		} else {
			char += data[i+1]
		}
		length := data[i+2]
		if char+length > UTF16Len(lines[line]) {
			t.Fatalf("token crosses its line: line %d char %d len %d", line, char, length)
		}
		if 1 == line && 0 == char {
			sawContinuation = true
		}
	}
	if !sawContinuation {
		t.Fatal("no continuation span on the second line")
	}
}
